README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# IMABC

<!-- badges: start -->
<!-- badges: end -->

## Installation

You can install the release version from CRAN with:

``` r
# install.packages("imabc")
```

You can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
# devtools::install_github("carolyner/imabc")
# library(imabc)
```

Other packages needed are: \* MASS \* data.table \* tidyverse \*
parallel \* foreach \* doParallel \* lhs \* truncnorm

## Running the Code

### Define Model Parameters/Priors

Parameters must have unique names (a unique name will be supplied if not
given)

``` r
priors <- define_priors(
  # x1: Uniform Prior (from base R)
  x1 = add_prior(
    dist_base_name = "unif",
    min = 0.2, max = 0.9
  ),
  # x2: Truncated Normal (from truncnorm package)
  add_prior(
    parameter_name = "x2",
    density_fn = "dtruncnorm",
    mean = 0.5, sd = 0.05, a = 0.4, b = 0.8, # a = min and b = max
    min = 0.4, max = 0.8 # User must specify both in truncnorm
  ),
  # V3: Fixed parameter (not calibrated)
  add_prior(0.5)
)
```

### Define Target Values

Targets must have unique names (a unique name will be supplied if not
given)

``` r
targets <- define_targets(
  # G1: Grouped targets include T1 and T2
  G1 = group_targets(
    T1 = add_target(
      target = 1.5,
      starting_range = c(1.0, 2.0),
      stopping_range = c(1.49, 1.51)
    ),
    add_target(
      target_name = "T2",
      target = 0.5,
      starting_range = c(0.2, 0.9),
      stopping_range = c(0.49, 0.51)
    )
  )
)
```

### Define Target Function

The target function must return a vector whose length is equal to the
number of targets. If the return vector is named using the target names
then the order of the vector will be corrected by `imabc()`. If the
return vector is not named, the targets should be added to the vector in
the same order they were added to the targets object. Finalize the
target function using the define\_target\_function to ensure the
function is specified correctly for `imabc()`.

``` r
fn1 <- function(x1, x2) { x1 + x2 + sample(c(-1, 1), 1)*rnorm(1, 0, 0.1) }
fn2 <- function(x1, x2) { x1 * x2 + sample(c(-1, 1), 1)*rnorm(1, 0, 0.1) }
fn <- function(x1, x2) {
  res <- c()
  res["T2"] <- fn1(x1, x2)
  res["T1"] <- fn2(x1, x2)
  return(res)
}
target_fun <- define_target_function(
  targets, priors, FUN = fn, use_seed = FALSE
)
```

### Calibrate Model

The primary function is imabc(). The inputs and their descriptions are
as follows:

``` r
calibration_results <- imabc(
  priors = priors,
  targets = targets_nogroup,
  target_fun = target_fun,
  seed = 54321,
  N_start = 2000,
  N_centers = 2,
  Center_n = 500,
  N_cov_points = 50,
  N_post = 100
)
```

-   target\_fun: A function that generate target values given parameters
    (i.e., \`the model’). The use of define\_target\_function is
    stronlgy advised to ensure that the function takes in the correct
    values and correctly returns results.
-   priors: A priors object created using define\_priors. This contains
    information regarding the parameters that are being calibrated. Is
    ignored if starting from previous results.
-   targets: A targets object created using define\_targets. This
    contains information regarding the target values which will be used
    to evaluate simulated parameters. Is ignored if starting from
    previous results.
-   N\_start: numeric(1). The number of draws to simulate for the first
    iteration.
    -   This number should be large relative to the number of parameters
        in order to ensure as much of the parameter space is explored as
        possible. Setting this number too low could result in the model
        getting stuck in a local minima. A good rule of thumb is to set
        this to 1000 x Number of Parameters
-   N\_centers: numeric(1). The number of centers to use for exploring
    the parameter space.
    -   See note with Center\_n.
-   Center\_n: numeric(1). The number of points to add around each
    center
    -   This work in tandem with N\_centers to define the number of
        centers around which new parameter vectors will be generated.
        During each iteration of imabc, the N\_centers best points will
        be selected from the current set of good draws. Then imabc will
        simulate Center\_n draws based on one of the N\_center points.
-   N\_cov\_points: numeric(1). The minimum number of points used to
    estimate the covariance matrix of valid parameters nearest each
    center point. The covariance matrix is used when simulating new
    parameter draws around the center. If 0 (default), uses 25\*number
    of parameters.
    -   Sets the minimum number of good parameters sets required in
        order to simulate new parameter vectors based on the covariance
        from the good parameter sets. If the calibration has not
        achieved this many good parameters, new parameters will be
        simulated around the N\_centers with the assumption that the
        parameters are independent of each other. This is also the
        minimum number of good parameter sets required before the model
        will attempt to narrow the current target bounds towards the
        stopping bounds. Setting N\_cov\_points too low can result in
        poor performance of the algorithm. A good rule of thumb is at
        least 10 x Number of Parameters.
-   N\_post: numeric(1). The weighted sample size that must be achieved
    using valid parameter values in order to stop algorithm.
    -   This is the final number of accepted parameter sets that produce
        target values within the stopping bound ranges required before
        the calibration will complete.
-   sample\_inflate: numeric(1). When generating new results for a given
    center, how many additional samples should be simulated to ensure
    enough valid (within range) parameters draws are simulated for the
    center.
-   max\_iter: numeric(1). The maximum number of iterations to attempt.
-   seed: numeric(1). The seed to set for reproducibility.
-   latinHypercube: logical(1). Should algorithm use a Latin Hypercube
    to generate first set of parameters.
-   backend\_fun: function. For advanced users only. Lets to user
    evaluate the target function(s) using their own backend, i.e.,
    simulate targets with an alternative parallel method. Only necessary
    if the backend method is not compatible with foreach. See details
    for requirements.
-   output\_directory: character(1). Path to save results to. If NULL
    (default), no results are saved. If a path is provided results are
    saved/updated every iteration. See details for more information.
-   output\_tag: character(1). Tag to add to result files names.
    “timestamp” (default) is a special code that adds the time and date
    the code was executed.
-   previous\_results\_dir: Optional character(1). Path to results
    stored during a previous run. If the user wishes to restart a run
    that didn’t complete the calibration, they can continue by using the
    outputs stored during the previous run.
-   previous\_results\_tag: Optional character(1). The tag that was
    added to the previous run output files.
-   verbose: logical(1). Prints out progress messages and additional
    information as the model works.
-   validate\_run: logical(1). If this is TRUE and an output\_directory
    is specified, the function will save all parameters generated by the
    model - even ones that were deemed invalid based on their simulated
    targets.

Some additional notes: \* Regarding parallelization: If the user wishes
to take advantage of parallelization they can register a parallel
backend before running the imabc function (e.g. registerDoParallel(cores
= detectCores() - 1)). imabc() uses foreach to submit parameters to the
target function so what works with it will work with imabc. If the
register function the user wishes to use requires that additional inputs
beyond the parameters data.table, the target function, and the special
other values listed in the backend\_fun description be passed to foreach
specifically the user must handle the entire submission of runs
themselves by defining the backend\_fun option in imabc. \* At the
moment, The distance between simulated targets and set target values
(desired targets) is calculated following a chisquare ((desired -
simulated)<sup>2</sup>)/(desired<sup>2</sup>)) except in the case where
the desired target is 0. For desired targets = 0, the function replaces
the denominator with 0.5 x range(stopping bounds):
((simulated)<sup>2</sup>)/((0.5 x (upper\_bound\_stop -
lower\_bound\_stop))<sup>2</sup>). If the user wishes to use the
stopping range to scale all the target distances they can set the global
option imabc.target\_eval\_distance to “stopping\_range” (i.e. before
running imabc, run
`options(imabc.target_eval_distance = "stopping_range")`)

## To do

-   Allow for continuing runs
    -   target modification
    -   parameter modification


NEWS.md
# imabc 1.0.1

* Added a `NEWS.md` file to track changes to the package.
* Fixes error where validate_runs = TRUE was not saving all the iterations for good_* files.


R/Priors.R
#' @name PriorsSpecification
#' @title Specify the Prior Distributions for All Parameters
#'
#' @param ... Optional. In add_prior: Named inputs to be passed to the RNG functions. In define_priors: The results of
#' add_prior calls - one for each parameter that is being calibrated.
#'
#' @description Helper functions that can be used to create an imabc priors object used by imabc().
#'
#' @section Distribution Specifications:
#' If the user does not provide any RNG functions specifications, they must provide a single value in order to
#' create a fixed parameter. This is not the most efficient method for using a fixed parameter in a model.
#'
#' If the user only provides one of the RNG functions specifications, these functions will search for the most logical
#' names for the other functions. I.e. if dist_base_name is provided (e.g. unif), these will assume that the user wishes
#' to use paste0("d", dist_base_name) for the density function and paste0("q", dist_base_name) for the quantile function.
#' These functions will make the corresponding guesses if the user provides density_fn or quantile_fn. If density_fn or
#' quantile_fn are provided, they will assume those functions are preferred over any calculated function names.
#' @md
#'
#' @section RNG Input Specifications:
#' These functions will attempt to pass any extra arguments to the RNG functions. These arguments must be named to match
#' the expected inputs not to create errors. If a value's name cannot be matched to an RNG function input, it will be
#' ignored.
#'
#' min/max are important values to imabc and will always be defined for each parameter. They are used to evaluate whether
#' any simulated parameters are valid. The user can specify values for them if they want. If the user does not specify
#' them they will look at the RNG function and if the RNG has default values for min/max it will use them, otherwise
#' it will use -Inf/Inf respectively. **Warning**: This behavior depends on the RNG functions using min and max as the input
#' names for the min and max values. If the RNG functions use an alternate name for these concepts they will treat
#' them as separate values. An example of this can be found in the truncnorm package which uses a and b for the min and
#' max respectively. For those functions the user would need to specify inputs for a, b, min, and max in order to get a
#' consistent result.
#' @md
#'
#' @section Parameter Names:
#' The user can specify names by either specifying the input parameter_name in add_prior or by setting the result of an
#' add_prior call to a object in define_priors (e.g. define_priors(x1 = add_prior(...))). If the user specifies the
#' parameter_name input and sets add_prior to an object, the parameter_name value will be used. If no name is specified
#' a unique name will be generated automatically.
#' @md
#'
#' @return A priors imabc object.
NULL

#' @rdname PriorsSpecification
#'
#' @param dist_base_name Optional character(1). The base name of the RNG function set (or the column with the dist_base_name
#' info in as.priors) for the prior distribution.
#' @param density_fn Optional character(1). The name of the RNG density function (or the column with the density_fn
#' info in as.priors) for the prior distribution.
#' @param quantile_fn Optional character(1). The name of the RNG quantile function (or the column with the quantile_fn
#' info in as.priors) for the prior distribution.
#' @param parameter_name Optional character(1). The name of the parameter (or the column with the parameter_name info in
#' as.priors).
#'
#' @examples
#' add_prior(dist_base_name = "norm")
#' add_prior(density_fn = "dnorm", mean = 50, sd = 10)
#' add_prior(quantile_fn = "qnorm", min = 0, max = 1)
#'
#' @export
add_prior <- function(..., dist_base_name = NULL, density_fn = NULL, quantile_fn = NULL, parameter_name = NULL) {
  inputs <- list(...)

  # if dist_base_name, density_fn, and quantile_fn are NULL, assume a fixed parameter
  if (is.null(dist_base_name) & is.null(density_fn) & is.null(quantile_fn)) {
    # Check appropriateness of input(s)
    stopifnot(
      "For a Fixed Parameter only supply a single numeric value of length 1." = length(inputs) == 1,
      "Input must be numeric." = !is.na(suppressWarnings(as.numeric(inputs)))
    )

    # Define outputs
    inputs <- as.numeric(inputs)
    density_function <- quantile_function <- function(n) {
      sims <- rep(inputs, length(n))

      return(sims)
    }
    # Define min/max for attributes
    min <- inputs
    max <- inputs
    distribution <- "fixed"
    fun_inputs <- inputs
  } else { # is.null(dist_base_name) & is.null(density_fn) & is.null(quantile_fn)
    if (is.null(dist_base_name)) {
      if (!is.null(density_fn) & is.null(quantile_fn)) {
        # dist_base_name  density_fn  quantile_fn
        #           NULL    provided         NULL
        # Check that density function provided exists
        density_fn <- .validate_prior_function(density_fn)
        # Check that best guess quantile function exists
        quantile_fn <- .validate_prior_function(gsub("^d", "q", density_fn))

      } else if (is.null(density_fn) & !is.null(quantile_fn)) {
        # dist_base_name  density_fn  quantile_fn
        #           NULL        NULL     provided
        # Check that quantile function provided exists
        quantile_fn <- .validate_prior_function(quantile_fn)
        # Check that best guess density function exists
        density_fn <- .validate_prior_function(gsub("^q", "d", quantile_fn))

      } else if (!is.null(density_fn) & !is.null(quantile_fn)) {
        # dist_base_name  density_fn  quantile_fn
        #           NULL    provided     provided
        # Check that density function provided exists
        density_fn <- .validate_prior_function(density_fn)
        # Check that quantile function provided exists
        quantile_fn <- .validate_prior_function(quantile_fn)
      }
      # Distribution shorthand
      dist_base_name <- gsub("^d", "", density_fn)

    } else if (!is.null(dist_base_name)) {
      # Check that base name converts to existing function
      density_fn_guess <- .validate_prior_function(paste0("d", dist_base_name))
      # Check that quantile function provided exists
      quantile_fn_guess <- .validate_prior_function(paste0("q", dist_base_name))

      if (is.null(density_fn) & is.null(quantile_fn)) {
        # dist_base_name  density_fn  quantile_fn
        #       provided        NULL         NULL
        density_fn <- density_fn_guess
        quantile_fn <- quantile_fn_guess

      } else if (!is.null(density_fn) & is.null(quantile_fn)) {
        # dist_base_name  density_fn  quantile_fn
        #       provided    provided         NULL
        density_fn <- .validate_prior_function(density_fn)
        # Warn if density function from dist_base_name != density_fn
        if (density_fn_guess != density_fn) {
          warning("Base distribution name does not convert to provided density function. Using density_fn.")
        }
        quantile_fn <- quantile_fn_guess

      } else if (is.null(density_fn) & !is.null(quantile_fn)) {
        # dist_base_name  density_fn  quantile_fn
        #       provided        NULL     provided
        quantile_fn <- .validate_prior_function(quantile_fn)
        # Warn if quantile function from dist_base_name != quantile_fn
        if (quantile_fn_guess != quantile_fn) {
          warning("Base distribution name does not convert to provided quantile function. Using quantile_fn.")
        }
        density_fn <- density_fn_guess

      } else if (!is.null(density_fn) & !is.null(quantile_fn)) {
        # dist_base_name  density_fn  quantile_fn
        #       provided    provided     provided
        density_fn <- .validate_prior_function(density_fn)
        quantile_fn <- .validate_prior_function(quantile_fn)

        # Warn if density function from dist_base_name != density_fn
        if (density_fn_guess != density_fn) {
          warning("Base distribution name does not convert to provided density function. Using density_fn.")
        }
        # Warn if quantile function from dist_base_name != quantile_fn
        if (quantile_fn_guess != quantile_fn) {
          warning("Base distribution name does not convert to provided quantile function. Using quantile_fn.")
        }

      }
    } # !is.null(dist_base_name)

    # Defaults for min and max values of prior distribution: inputs -> formals -> -Inf/Inf
    min <- ifelse(
      "min" %in% names(inputs), inputs$min, ifelse(
        "min" %in% names(formals(density_fn)), formals(density_fn)["min"], -Inf
      ))
    max <- ifelse("max" %in% names(inputs), inputs$max, ifelse(
      "max" %in% names(formals(density_fn)), formals(density_fn)["max"], Inf
    ))

    # Define the density function call
    density_inputs <- inputs[intersect(names(inputs), formalArgs(density_fn)[-1])]
    density_function <- function(n) {
      sims <- mapply(match.fun(density_fn), n, MoreArgs = density_inputs)

      return(sims)
    }

    # Define the density function call
    quantile_inputs <- inputs[intersect(names(inputs), formalArgs(density_fn)[-1])]
    quantile_function <- function(n) {
      sims <- mapply(match.fun(quantile_fn), n, MoreArgs = quantile_inputs)

      return(sims)
    }

    distribution <- dist_base_name
    fun_inputs <- unlist(quantile_inputs)
  } # ! is.null(dist_base_name) & is.null(density_fn) & is.null(quantile_fn)

  # Create the prior object
  prior <- structure(
    list(
      parameter = parameter_name,
      density_function = density_function,
      quantile_function = quantile_function,
      min = min,
      max = max,
      distribution = distribution,
      fun_inputs = fun_inputs
    ),
    class = c("prior", "imabc")
  )

  return(prior)
}

#' @rdname PriorsSpecification
#'
#' @param prior_df Optional data.frame. Priors stored as a data.frame or from the results object of a previous run.
#'
#' @examples
#' # x1, x2, and x3 reflect three parameters in the mdoel.
#' x1 <- add_prior(dist_base_name = "norm")
#' define_priors(
#'   x1 = x1,
#'   x2 = add_prior(density_fn = "dnorm", mean = 50, sd = 10),
#'   add_prior(parameter_name = "x3", quantile_fn = "qnorm", min = 0, max = 1)
#' )
#'
#' @export
define_priors <- function(..., prior_df = NULL) {
  priors_list <- list(...)
  if (length(priors_list) > 0 & !is.null(prior_df)) {
    stop("Currently does not support adding new parameters to an old set of runs")
  }

  # When new priors are defined
  if (length(priors_list) > 0) {
    # Check that only imabc_priors have been added
    good <- unlist(lapply(priors_list, function(x) { all(inherits(x, c("prior", "imabc"), TRUE) != 0) }))
    if (!all(good)) {
      bad_noname <- which(names(good) == "" & !good)
      bad_winame <- which(names(good) != "" & !good)
      txt1 <- txt2 <- c()
      if (length(bad_noname) > 0) {
        txt1 <- paste("prior's at position(s)", paste(bad_noname, collapse = ", "), sep = ": ")
      }
      if (length(bad_winame) > 0) {
        txt2 <- paste("prior's named", paste(names(bad_winame), collapse = ", "), sep = ": ")
      }
      if (!is.null(txt1) && !is.null(txt2)) {
        txt <- paste(txt1, "and", txt2, sep = " ")
      } else if (!is.null(txt1)) {
        txt <- txt1
      } else if (!is.null(txt2)) {
        txt <- txt2
      }
      stop(sprintf("Inputs must be imabc priors. Use add_prior() to create prior objects for %s.", txt))
    }

    # Store added functions
    return_priors <- get_list_element(priors_list, name = c("density_function", "quantile_function"), unlist = FALSE)
    return_priors <- structure(return_priors, class = c("priors", "imabc"))

    # Pull all important attributes of the priors into lists
    name_vec <- get_list_element(priors_list, "parameter", unlist = TRUE)
    sd <- rep.int(0L, length(priors_list)) # empirically calculated in imabc
    min <- get_list_element(priors_list, "min", unlist = TRUE) # if not set or given default by distribution functions, defaults to -Inf
    max <- get_list_element(priors_list, "max", unlist = TRUE) # if not set or given default by distribution functions, defaults to Inf
    distribution <- get_list_element(priors_list, "distribution", unlist = TRUE)
    fun_inputs <- get_list_element(priors_list, "fun_inputs", unlist = FALSE)

    # Pull the names of the parameters. Check for uniqueness and add unique names where no names are provided.
    prior_names <- unique_names(return_priors, name_vec)
    names(return_priors) <- prior_names
    names(min) <- prior_names
    names(max) <- prior_names
    names(sd) <- prior_names
    names(distribution) <- prior_names
    names(fun_inputs) <- prior_names

    # Store the vectors as attributes for the entire prior object
    attributes(return_priors)$mins <- min
    attributes(return_priors)$maxs <- max
    attributes(return_priors)$sds <- sd
    attributes(return_priors)$distributions <- distribution
    attributes(return_priors)$fun_inputs <- fun_inputs
  }

  # If reading from a data.frame or previous imabc run set of results
  if (!is.null(prior_df)) {
    return_priors <- as.priors(prior_df)
  }

  return(return_priors)
}

#' @rdname PriorsSpecification
#'
#' @param df data.frame. Each parameter should be a row and each column is an input into add_prior. If a given column doesn't
#' relate to a parameter, set its value to NA.
#'
#' @examples
#' df <- data.frame(
#'   parameter_name = c("x1", "x2", "x3"), dist_base_name = c("unif", NA, NA), density_fn = c(NA, "dtruncnorm", NA),
#'   quantile_fn = c(NA, NA, "qnorm"), mean = c(NA, 0.75, NA), sd = c(NA, 0.05, NA), min = c(x1_min, x2_min, NA),
#'   max = c(x1_max, x2_max, NA), a = c(NA, x2_min, NA), b = c(NA, x2_max, NA)
#' )
#' as.priors(df, name_var = "parameter_name", dist_var = "dist_base_name", density_var = "density_fn", quantile_var = "quantile_fn")
#'
#' @export
as.priors <- function(df, ...) {
  # Columns: parameter_name = NULL, dist_base_name = NULL, density_fn = NULL, quantile_fn = NULL
  col_names_alt <- do.call(c, list(...))
  col_names_dta <- colnames(df)

  # Check that at least one distribution function name column exists
  if (!any(c("dist_base_name", "density_fn", "quantile_fn") %in% c(col_names_dta, names(col_names_alt)))) {
    stop(paste(
      "Function name column(s) not found in data. Add at least one column named dist_base_name, density_fn, or quantile_fn. ",
      "You can also point as.priors to the appropriate column like so:\n",
      'as.priors(df, dist_base_name = "alternate name")\n',
      "See ?add_prior to understand what each column should represent.", sep = ""
    ))
  }

  # If any alternate names are provided, rename the columns appropriately
  if (length(col_names_alt) > 0) {
    change_name_idx <- match(col_names_alt, col_names_dta)
    not_found_cols <- is.na(change_name_idx)
    if (any(not_found_cols)) {
      warning(sprintf("Ignoring unknown column name: %s\n", names(col_names_alt)[not_found_cols]))
      change_name_idx <- change_name_idx[!not_found_cols]
      col_names_alt <- col_names_alt[!not_found_cols]
    }
    col_names_dta[change_name_idx] <- names(col_names_alt)
  }

  # Make sure the names we can control match the expected name for add_prior()
  colnames(df) <- col_names_dta

  # For each parameter, create a prior object
  prio_list <- lapply(1:nrow(df), FUN = function(i1, dta) {
    tmp <- dta[i1, , drop = TRUE]
    tmp <- tmp[!is.na(tmp)]

    # Make sure fixed distributions are converted to single scalar value
    if (any(tmp[c("dist_base_name", "density_fn", "quantile_fn")] == "fixed")) {
      # Remove distribution objects
      tmp[c("dist_base_name", "density_fn", "quantile_fn")] <- NULL

      # Join parameter_name if it exits with first non-NA value you find
      new_tmp <- tmp[names(tmp) == "parameter_name"]
      new_tmp <- c(new_tmp, tmp[names(tmp) != "parameter_name"][[1]])
      tmp <- new_tmp
    }

    do.call(add_prior, tmp)
  }, dta = df)

  # Take all prior objects and add them into a final imabc priors object
  final_prio_list <- do.call(define_priors, prio_list)

  # Add empirical sds if they exist
  if ("prior_sd" %in% col_names_dta) {
    # Get sds in proper order with names
    parms <- names(final_prio_list)
    sds <- df[["prior_sd"]][match(parms, df[["parameter_name"]])]
    names(sds) <- parms
    attr(final_prio_list, "sds") <- sds
  }

  return(final_prio_list)
}

#' @export
as.data.frame.priors <- function(priors_list) {
  # Put non-function specific inputs into data.frame
  parameter_name <- names(priors_list)
  dist_base_name <- attr(priors_list, "distributions")
  min <- attr(priors_list, "mins")
  max <- attr(priors_list, "maxs")
  prior_sd <- attr(priors_list, "sds")
  out_df <- data.frame(parameter_name, dist_base_name, min, max, prior_sd)
  rownames(out_df) <- NULL
  # Remove empirical SD if it hasn't been calculated yet
  if (all(out_df$prior_sd == 0)) { out_df$prior_sd <- NULL }

  # Turn function specific inputs into a data.frame
  fn_inputs <- lapply(attr(priors_list, "fun_inputs"), FUN = function(x) {
    fn_in <- names(x$fun_inputs)
    df_out <- NA
    if (!is.null(fn_in)) {
      df_out <- as.data.frame(t(x$fun_inputs))
      df_out$min <- NULL
      df_out$max <- NULL

      if (ncol(df_out) == 0) {
        df_out <- NA
      }
    }

    return(df_out)
  })
  fn_inputs <- do.call(rbind, fn_inputs)

  # Make sure at least one column of inputs has inputs
  if (any(!is.na(fn_inputs[, 1]))) {
    fn_inputs <- cbind(rownames(fn_inputs), fn_inputs)
    fn_inputs <- as.data.frame(fn_inputs)
    rownames(fn_inputs) <- NULL
    colnames(fn_inputs)[1] <- "parameter_name"

    # Merge non-function specific inputs with function specific inputs
    out_df <- merge(out_df, fn_inputs, by = "parameter_name", sort = FALSE)
  }

  return(out_df)
}

#' @export
`[.priors` <- function(p, x) {
  old_class <- class(p)

  # Handle if subset is a list of names or T/F
  if (is.logical(x)) {
    keep <- x
  } else {
    keep <- names(p) %in% x
  }

  # Add subset of target info to a new target lsit
  subset_priors <- structure(
    unclass(p)[keep],
    class = old_class
  )

  # Subset attributes
  mins <- attributes(p)$mins[keep]
  maxs <- attributes(p)$maxs[keep]
  sds <- attributes(p)$sds[keep]
  distributions <- attributes(p)$distributions[keep]
  fun_inputs <- attributes(p)$fun_inputs[keep]
  # Add them to new target list
  attributes(subset_priors)$mins <- mins
  attributes(subset_priors)$maxs <- maxs
  attributes(subset_priors)$sds <- sds
  attributes(subset_priors)$distributions <- distributions
  attributes(subset_priors)$fun_inputs <- fun_inputs

  return(subset_priors)
}

#' @export
print.priors <- function(p, digits = getOption("digits")) {
  n_parms <- length(p)
  parm_names <- names(p)

  cat(sprintf("There are %s defined parameters.\n", n_parms))
  p_text <- lapply(parm_names, FUN = function(x, p) {
    dist_text <- sprintf("Distribution base name: %s", attr(p, "distributions")[x])
    fun_text <- sprintf("User specified inputs: %s", attr(p, "fun_inputs")[x])
    fun_text <- sub("(.*)(list\\(fun_inputs = c*\\(*)(.*)(\\))", "\\1\\3", fun_text, perl = T)
    fun_text <- sub("\\)", "", fun_text)
    range_text <- sprintf("Allowable range: %s - %s",
                          format(attr(p, "mins")[x], digits = digits),
                          format(attr(p, "maxs")[x], digits = digits))
    sd_text <- sprintf("Empirical Standard Deviation: %s", format(attr(p, "sds")[x], digits = digits))

    full_text <- paste(dist_text, fun_text, range_text, sd_text, sep = "\n")
  }, p = p)
  cat(
    paste(sprintf("\nParameter %s has the following specifications:", parm_names), p_text, sep = "\n", collapse = "\n")
  )

  invisible(p)
}
